src/main/cpp/DragonVision/DragonVision.cpp

```diff
@@ -277,121 +277,160 @@ std::optional<VisionData> DragonVision::GetVisionDataFromNote(VISION_ELEMENT ele
 		break;
 	case VISION_ELEMENT::SOURCE:
-		fieldElementPose = allianceColor == frc::DriverStation::Alliance::kRed ? frc::Pose3d{m_constants->GetFieldElement(FieldConstants::FIELD_ELEMENT::RED_SOURCE)} /*load red source*/ : frc::Pose3d{FieldConstants::GetInstance()->GetFieldElement(FieldConstants::FIELD_ELEMENT::BLUE_SOURCE)}; /*load blue source*/
+		fieldElementPose = allianceColor == frc::DriverStation::Alliance::kRed ? frc::Pose3d{FieldConstants::GetInstance()->GetFieldElement(FieldConstants::FIELD_ELEMENT::RED_SOURCE)} /*load red source*/ : frc::Pose3d{FieldConstants::GetInstance()->GetFieldElement(FieldConstants::FIELD_ELEMENT::BLUE_SOURCE)}; /*load blue source*/
 		break;
 	default:
-		// no-op
+		return std::nullopt;
 		break;
 	}
 
-	// double check selectedCam is not nullptr
-	if (selectedCam != nullptr)
+	std::optional<VisionData> multiTagEstimate = MultiTagToElement(fieldElementPose);
+	std::optional<VisionData> singleTagEstimate = SingleTagToElement(fieldElementPose);
+	if (multiTagEstimate)
 	{
+		return multiTagEstimate;
+	}
+	else if (singleTagEstimate)
+	{
+		return singleTagEstimate;
+	}
 
-		// check for a more accurate multitag estimate
-		std::optional<VisionPose> multitagPose = static_cast<DragonPhotonCam *>(selectedCam)->GetMultiTagEstimate();
-		if (multitagPose)
-		{
-			// get robot pose from multitag estimate
-			frc::Pose3d robotPose = multitagPose.value().estimatedPose;
+	return std::nullopt;
+}
+
+std::optional<VisionData> DragonVision::MultiTagToElement(frc::Pose3d elementPose)
+{
+	// MULTITAG
+	// std::optional<VisionPose> launcherMultiTag = std::nullopt;
+	// std::optional<VisionPose> placerMultiTag = std::nullopt;
+	/*if (m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::LAUNCHER] != nullptr)
+	{
+		launcherMultiTag = dynamic_cast<DragonPhotonCam *>(m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::LAUNCHER])->GetMultiTagEstimate();
+	}
 
-			// calculate transform to fieldElement as difference between robot pose and field element pose
-			frc::Transform3d transformToElement = frc::Transform3d{robotPose, fieldElementPose};
+	if (m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::PLACER] != nullptr)
+	{
+		placerMultiTag = dynamic_cast<DragonPhotonCam *>(m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::LAUNCHER])->GetMultiTagEstimate();
+	}*/
 
-			// calculate rotation3d for angles from robot center, not transformation
-			units::angle::radian_t pitch = units::math::atan2(transformToElement.Z(), transformToElement.X());
-			units::angle::radian_t yaw = units::math::atan2(transformToElement.Y(), transformToElement.X());
-			units::angle::radian_t roll = units::math::atan2(transformToElement.Z(), transformToElement.Y());
-			frc::Rotation3d rotation = frc::Rotation3d(roll, pitch, yaw);
+	std::optional<VisionPose> launcherMultiTag = dynamic_cast<DragonPhotonCam *>(m_dragonCameraMap[RobotElementNames::CAMERA_USAGE::LAUNCHER])->GetMultiTagEstimate();
```

Comment:
we may want to make sure m_dragonCameraMap{RobotElementNames::xxx] is not = null before we cast it and read from it. This would be in cases where a robot didn't have phonsiion, but was still calling this code

Reply:
Done